test/common/config/delta_subscription_state_test.cc

```diff
@@ -26,22 +28,63 @@ namespace Config {
 namespace {
 
 const char TypeUrl[] = "type.googleapis.com/envoy.api.v2.Cluster";
+enum class LegacyOrUnified { Legacy, Unified };
 
-class DeltaSubscriptionStateTestBase : public testing::Test {
+class DeltaSubscriptionStateTestBase : public testing::TestWithParam<LegacyOrUnified> {
 protected:
   DeltaSubscriptionStateTestBase(
-      const std::string& type_url, const bool wildcard,
-      const absl::flat_hash_set<std::string> initial_resources = {"name1", "name2", "name3"})
-      : timer_(new Event::MockTimer(&dispatcher_)),
-        state_(type_url, callbacks_, local_info_, dispatcher_, wildcard) {
-    state_.updateSubscriptionInterest(initial_resources, {});
-    envoy::service::discovery::v3::DeltaDiscoveryRequest cur_request =
-        state_.getNextRequestAckless();
-    EXPECT_THAT(cur_request.resource_names_subscribe(),
+      const std::string& type_url, const bool wildcard, LegacyOrUnified legacy_or_unified,
+      const absl::flat_hash_set<std::string> initial_resources = {"name1", "name2", "name3"},
+      const unsigned int initial_fetch_timeout = 0U)
+      : should_use_unified_(legacy_or_unified == LegacyOrUnified::Unified) {
+    if (should_use_unified_ && initial_fetch_timeout > 0) {
+      initial_fetch_timeout_timer_ = new Event::MockTimer(&dispatcher_);
+      EXPECT_CALL(*initial_fetch_timeout_timer_,
+                  enableTimer(std::chrono::milliseconds(initial_fetch_timeout), _));
+    }
+
+    ttl_timer_ = new Event::MockTimer(&dispatcher_);
+
+    if (should_use_unified_) {
+      state_ = std::make_unique<Envoy::Config::UnifiedMux::DeltaSubscriptionState>(
+          type_url, callbacks_, std::chrono::milliseconds(initial_fetch_timeout), dispatcher_,
+          wildcard);
+    } else {
+      state_ = std::make_unique<Envoy::Config::DeltaSubscriptionState>(
+          type_url, callbacks_, local_info_, dispatcher_, wildcard);
+    }
+    updateSubscriptionInterest(initial_resources, {});
+    auto cur_request = getNextRequestAckless();
+    EXPECT_THAT(cur_request->resource_names_subscribe(),
                 // UnorderedElementsAre("name1", "name2", "name3"));
                 UnorderedElementsAreArray(initial_resources.cbegin(), initial_resources.cend()));
   }
 
+  void updateSubscriptionInterest(const absl::flat_hash_set<std::string>& cur_added,
+                                  const absl::flat_hash_set<std::string>& cur_removed) {
+    if (should_use_unified_) {
+      absl::get<1>(state_)->updateSubscriptionInterest(cur_added, cur_removed);
```

Comment:
Can you use subclassing here to simplify this branching logic everywhere?

Reply:
Not easily: state classes don't have a common interface, so I'd need to introduce one, and then create wrapper classes for two concrete state classes (the old and the new implementation of delta_state). At which point it's probably going to be more trouble than it's worth?

Reply:
Ack - presumably this all goes away when we have removed the old implementation anyway.